packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/timestamp.dart

```diff
@@ -40,8 +40,15 @@ class Timestamp implements Comparable<Timestamp> {
 
   /// Create a [Timestamp] fromMicrosecondsSinceEpoch
   factory Timestamp.fromMicrosecondsSinceEpoch(int microseconds) {
-    final int seconds = microseconds ~/ _kMillion;
-    final int nanoseconds = (microseconds - seconds * _kMillion) * _kThousand;
+    int seconds = microseconds ~/ _kMillion;
+    int nanoseconds = (microseconds - seconds * _kMillion) * _kThousand;
+
+    // Matches implementation in Android SDK:
+    // https://github.com/firebase/firebase-android-sdk/blob/master/firebase-common/src/main/java/com/google/firebase/Timestamp.kt#L114-L121
+    if (nanoseconds < 0) {
+      seconds -= 1;
+      nanoseconds += _kBillion;
+    }
```

Comment:
@SelaseKay
Also the factory `fromMillisecondsSinceEpoch` should be fixed.
I personally would delegate the milliseconds one to the microseconds one:
```dart
factory Timestamp.fromMillisecondsSinceEpoch(int milliseconds) {
  return fromMicrosecondsSinceEpoch(milliseconds * _kThousand);
}
```
but I don't know if there are other considerations to be done